File: OutputWindow/cShaderData.cpp
#include "cShaderData.h"

void ShaderData::SetSource(std::string name, std::string version, std::string source)
{
    sName = name;
    sVersion = version;
    sSource = source;
}

void ShaderData::ClearLayout()
{
    rawLayout.clear();
}

void ShaderData::AddLayout(D3D11_INPUT_ELEMENT_DESC element)
{
    rawLayout.push_back(element);
}

bool ShaderData::CompileShaderFromString(ID3D11Device* dev)
{
    HRESULT result = S_OK;
    ID3D10Blob* errorMessage;
    ID3D10Blob* shader;

    DWORD flags = D3DCOMPILE_PACK_MATRIX_COLUMN_MAJOR | D3DCOMPILE_ENABLE_STRICTNESS | D3DCOMPILE_WARNINGS_ARE_ERRORS;
#ifdef _DEBUG
    flags |= D3DCOMPILE_SKIP_OPTIMIZATION | D3DCOMPILE_DEBUG;
#else
    flags |= D3DCOMPILE_OPTIMIZATION_LEVEL3;
#endif

    result = D3DCompile(sSource.c_str(), strlen(sSource.c_str()), 0, 0, 0, sName.c_str(), sVersion.c_str(), flags, 0, &shader, &errorMessage);
    if (FAILED(result)) {
        if (errorMessage) {
            std::string errorMsg = "Error compiling shader '" + sName + "'";
            MessageBoxA(0, errorMsg.c_str(), "CompileShaderFromString", MB_OK);
            return false;
        }
        else {
            std::string errorMsg = "Error compiling shader '" + sName + "'";
            MessageBoxA(0, errorMsg.c_str(), "CompileShaderFromString", MB_OK);
            return false;
        }
    }

    result = S_OK;
    if (sVersion.substr(0, 3) == "vs_")
    {
        result = dev->CreateInputLayout(&rawLayout[0], rawLayout.size(), shader->GetBufferPointer(), shader->GetBufferSize(), &Layout);
        if (FAILED(result)) {
            std::string errorMsg = "Error creating layout for shader '" + sName + "'";
            MessageBoxA(0, errorMsg.c_str(), "CompileShaderFromString", MB_OK);
            return false;
        }

        result = dev->CreateVertexShader(shader->GetBufferPointer(), shader->GetBufferSize(), NULL, &VS);
    }
    else if (sVersion.substr(0, 3) == "ps_")
        result = dev->CreatePixelShader(shader->GetBufferPointer(), shader->GetBufferSize(), NULL, &PS);

    if (FAILED(result)) {
        std::string errorMsg = "Error creating shader '" + sName + "'";
        MessageBoxA(0, errorMsg.c_str(), "CompileShaderFromString", MB_OK);
        return false;
    }

    shader->Release();
    shader = nullptr;

    return true;
}

void ShaderData::Release()
{
    if (VS) { VS->Release(); VS = nullptr; }
    if (PS) { PS->Release(); PS = nullptr; }
    if (Layout) { Layout->Release(); Layout = nullptr; }
}

File: MaskedCarnivale/DirectX/ShaderData/cShaderData.cs
using System;
using System.Collections.Generic;

using SharpDX.D3DCompiler;
using Device11 = SharpDX.Direct3D11.Device5;
using InputElement = SharpDX.Direct3D11.InputElement;
using VertexShader = SharpDX.Direct3D11.VertexShader;
using PixelShader = SharpDX.Direct3D11.PixelShader;
using InputLayout = SharpDX.Direct3D11.InputLayout;

namespace MaskedCarnivale.DirectX.ShaderData;
public class ShaderData
{
    string sName = "";
    string sVersion = "";
    string sSource = "";
    List<InputElement> rawLayout = new List<InputElement>();

    public VertexShader? VS { get; private set; } = null;
    public PixelShader? PS { get; private set; } = null;
    public InputLayout? Layout { get; private set; } = null;

    public void SetSource(string name, string version, string source)
    {
        sName = name;
        sVersion = version;
        sSource = source;
    }

    public void ClearLayout()
    {
        rawLayout.Clear();
    }

    public void AddLayout(InputElement element)
    {
        rawLayout.Add(element);
    }

    public bool CompileShaderFromString(Device11 dev)
    {
        ShaderFlags flags = ShaderFlags.PackMatrixColumnMajor | ShaderFlags.EnableStrictness | ShaderFlags.WarningsAreErrors;
#if DEBUG
        flags |= ShaderFlags.SkipOptimization | ShaderFlags.Debug;
#else
        flags |= ShaderFlags.OptimizationLevel3;
#endif
        ShaderBytecode shaderByteCode;
        try
        {
            shaderByteCode = ShaderBytecode.Compile(sSource, sName, sVersion, flags);
        }
        catch (Exception ex)
        {
            Plugin.Log!.Error($"Error compiling shader '{sName}' : {ex}");
            return false;
        }

        if (sVersion.Substring(0, 3) == "vs_")
        {
            try
            {
                Layout = new InputLayout(dev, shaderByteCode, rawLayout.ToArray());
            }
            catch (Exception ex)
            {
                Plugin.Log!.Error($"Error creating layout for shader '{sName}' : {ex}");
                return false;
            }

            VS = new VertexShader(dev, shaderByteCode);
        }
        else if (sVersion.Substring(0, 3) == "ps_")
            PS = new PixelShader(dev, shaderByteCode);

        shaderByteCode?.Dispose();
        return true;
    }

    public void Release()
    {
        VS?.Dispose();
        PS?.Dispose();
        Layout?.Dispose();
    }
}
